### ISSUE TITLE
Bug on Stack up exec

### ISSUE THREAD

**Issue by Manifestus** [pulumi/pulumi#17561]
### What happened?

```
error: The Pulumi CLI encountered a snapshot integrity error. This is a bug!

================================================================================
We would appreciate a report: https://github.com/pulumi/pulumi/issues/
Please provide all of the below text in your report.
================================================================================
Pulumi Version:    v3.130.0
Go Version:        go1.22.6
Go Compiler:       gc
Architecture:      arm64
Operating System:  darwin
Command:           pulumi up -s dev
Error:             resource urn:pulumi:dev::ss-pulumi::aws:iam/rolePolicyAttachment:RolePolicyAttachment::ss-pipe-dev-ecs-execution-role-policy-cloudwatch's dependency urn:pulumi:dev::ss-pulumi::aws:iam/role:Role::ss-pipe-dev-ecs-execution-role refers to missing resource

Stack Trace:

goroutine 1 [running]:
runtime/debug.Stack()
        /opt/hostedtoolcache/go/1.22.6/x64/src/runtime/debug/stack.go:24 +0x64
github.com/pulumi/pulumi/pkg/v3/resource/deploy.SnapshotIntegrityErrorf({0x106cbe376?, 0x4c?}, {0x1400248d048?, 0x4c?, 0x40?})
        /home/runner/work/pulumi/pulumi/pkg/resource/deploy/snapshot.go:313 +0x30
github.com/pulumi/pulumi/pkg/v3/resource/deploy.(*Snapshot).VerifyIntegrity(0x14002596900)
        /home/runner/work/pulumi/pulumi/pkg/resource/deploy/snapshot.go:217 +0xd9c
github.com/pulumi/pulumi/pkg/v3/backend/httpstate.(*cloudBackend).getSnapshot(0x60?, {0x107eb98c8, 0x1095bcd80}, {0x107e91720, 0x1095bcd80}, {0x107eb9de0?, 0x1400233e940?})
        /home/runner/work/pulumi/pulumi/pkg/backend/httpstate/state.go:229 +0x74
github.com/pulumi/pulumi/pkg/v3/backend/httpstate.(*cloudBackend).getTarget(0x1400208a0f0, {0x107eb98c8, 0x1095bcd80}, {0x107e91720, 0x1095bcd80}, {0x107eb9de0, 0x1400233e940}, 0x1400217d230, {0x107ea4240, 0x1400217d470})
        /home/runner/work/pulumi/pulumi/pkg/backend/httpstate/state.go:245 +0xa4
github.com/pulumi/pulumi/pkg/v3/backend/httpstate.(*cloudBackend).newUpdate(_, {_, _}, {_, _}, {0x1400214a160, {0x14000a394c0, 0x34}, {0x0, 0x0, ...}, ...}, ...)
        /home/runner/work/pulumi/pulumi/pkg/backend/httpstate/state.go:196 +0x1b0
github.com/pulumi/pulumi/pkg/v3/backend/httpstate.(*cloudBackend).runEngineAction(_, {_, _}, {_, _}, {_, _}, {0x1400214a160, {0x14000a394c0, 0x34}, ...}, ...)
        /home/runner/work/pulumi/pulumi/pkg/backend/httpstate/backend.go:1354 +0xf8
github.com/pulumi/pulumi/pkg/v3/backend/httpstate.(*cloudBackend).apply(_, {_, _}, {_, _}, {_, _}, {0x1400214a160, {0x14000a394c0, 0x34}, ...}, ...)
        /home/runner/work/pulumi/pulumi/pkg/backend/httpstate/backend.go:1328 +0x3ec
github.com/pulumi/pulumi/pkg/v3/backend.PreviewThenPrompt({_, _}, {_, _}, {_, _}, {0x1400214a160, {0x14000a394c0, 0x34}, {0x0, ...}, ...}, ...)
        /home/runner/work/pulumi/pulumi/pkg/backend/apply.go:120 +0x110
github.com/pulumi/pulumi/pkg/v3/backend.PreviewThenPromptThenExecute({_, _}, {_, _}, {_, _}, {0x1400214a160, {0x14000a394c0, 0x34}, {0x0, ...}, ...}, ...)
        /home/runner/work/pulumi/pulumi/pkg/backend/apply.go:247 +0xbc
github.com/pulumi/pulumi/pkg/v3/backend/httpstate.(*cloudBackend).Update(_, {_, _}, {_, _}, {0x1400214a160, {0x14000a394c0, 0x34}, {0x0, 0x0, ...}, ...})
        /home/runner/work/pulumi/pulumi/pkg/backend/httpstate/backend.go:1103 +0x90
github.com/pulumi/pulumi/pkg/v3/backend.UpdateStack({_, _}, {_, _}, {0x1400214a160, {0x14000a394c0, 0x34}, {0x0, 0x0, 0x0}, ...})
        /home/runner/work/pulumi/pulumi/pkg/backend/stack.go:96 +0x80
github.com/pulumi/pulumi/pkg/v3/backend/httpstate.(*cloudStack).Update(_, {_, _}, {0x1400214a160, {0x14000a394c0, 0x34}, {0x0, 0x0, 0x0}, 0x14000d95500, ...})
        /home/runner/work/pulumi/pulumi/pkg/backend/httpstate/stack.go:179 +0x64
main.newUpCmd.func1({_, _}, {{{0x1095bcd80, 0x0, 0x0}, {0x0, 0x0, 0x0}, 0x28, 0x0, ...}, ...}, ...)
        /home/runner/work/pulumi/pulumi/pkg/cmd/pulumi/up.go:187 +0xa74
main.newUpCmd.func3(0x14001f98908, {0x14001fbc4c0, 0x0, 0x0?})
        /home/runner/work/pulumi/pulumi/pkg/cmd/pulumi/up.go:531 +0x7f8
main.newUpCmd.runCmdFunc.func4(0x0?, {0x14001fbc4c0?, 0x0?, 0x0?})
        /home/runner/work/pulumi/pulumi/pkg/cmd/pulumi/errors.go:31 +0x28
main.newUpCmd.runCmdFunc.RunFunc.func7(0x14001f98908, {0x14001fbc4c0, 0x0, 0x2})
        /home/runner/work/pulumi/pulumi/sdk/go/common/util/cmdutil/exit.go:99 +0x38
github.com/spf13/cobra.(*Command).execute(0x14001f98908, {0x14001fbc4a0, 0x2, 0x2})
        /home/runner/go/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:987 +0x828
github.com/spf13/cobra.(*Command).ExecuteC(0x14001313b08)
        /home/runner/go/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:1115 +0x344
github.com/spf13/cobra.(*Command).Execute(...)
        /home/runner/go/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:1039
main.main()
        /home/runner/work/pulumi/pulumi/pkg/cmd/pulumi/main.go:55 +0x54
```

### Example

I just used pulumi up -s dev, this is on a different version of pulumi, 1 update behind

### Output of `pulumi about`

manifestus@Victors-MacBook-Pro ss-pulumi % pulumi about
CLI          
Version      3.130.0
Go Version   go1.22.6
Go Compiler  gc

Plugins
KIND      NAME    VERSION
resource  aws     6.49.1
resource  awsx    2.14.0
resource  docker  4.5.5
resource  docker  3.6.1
language  nodejs  unknown

Host     
OS       darwin
Version  15.0.1
Arch     arm64

This project is written in nodejs: executable='/Users/manifestus/.nvm/versions/node/v22.7.0/bin/node' version='v22.7.0'

Backend        
Name           pulumi.com
URL            https://app.pulumi.com/Manifestus
User           Manifestus
Organizations  Manifestus, skillstat-dev
Token type     personal

Dependencies:
NAME            VERSION
@pulumi/aws     6.49.1
@pulumi/awsx    2.14.0
@pulumi/pulumi  3.129.0
@types/node     18.19.44
typescript      5.5.4

Pulumi locates its logs in /var/folders/6d/n5y85v5x401d_h823g08gp5c0000gn/T/ by default
warning: Failed to get information about the current stack: No current stack

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by lunaris** [comment]
Hi @Manifestus, thanks for raising this and sorry you've hit an issue. I notice that you are on 3.130.0 -- do you know if the resource in question/any of those it references would have been using `deleteBeforeReplace`? If so then I think there's a good chance you might have hit https://github.com/pulumi/pulumi/pull/17146, which was fixed in 3.131.0. If not then let us know and we can hopefully find a way to find more details and resolve this.

In terms of unblocking you, I would recommend that you upgrade to the latest version and try the new `pulumi state repair` command to repair your state if you have not already done so by manual means. Additionally, later versions should hopefully contain fewer bugs that lead to errors like this.

**Comment by lunaris** [comment]
Hi @Manifestus, I'm going to close this as we haven't heard anything back -- if you have more details that could help us reproduce us, please don't hesitate to reopen this issue!
